src/AppBundle/Service/File/Storage/S3Storage.php

```diff
@@ -101,14 +102,71 @@ public function delete($key)
      * Remove an object and all its versions from S3 completely.
      *
      * @param  string      $key
-     * @return \Aws\Result
+     * @return array
      */
     public function removeFromS3($key)
     {
-        return $this->s3Client->deleteObject([
-            'Bucket' => $this->bucketName,
-            'Key'    => $key
-        ]);
+        if (!empty($key)) {
+            /*
+             * ListObjectVersions is permitted by ListBucketVersions in IAM.
+             */
+            $objectVersions = $this->s3Client->listObjectVersions([
+                'Bucket' => $this->bucketName,
+                'Prefix' => $key
+            ]);
+
+            $objectResult = [];
+            $objectsToDelete = '';
+            if ($objectVersions instanceof ResultInterface) {
```

Comment:
#notablocker It's usually better to test for the negative and return/throw early, to avoid large chunks of nested code. i.e:

```
if (!$objectVersions instanceof ResultInterface) {
    throw new \RuntimeException('Could not remove file: No results returned');
}
```

Reply:
Revised